xcodebuild.go

```diff
@@ -209,70 +224,104 @@ func runTest(buildTestParams models.XcodeBuildTestParamsModel, outputTool, xcpre
 	}
 
 	xcodebuildArgs = append(xcodebuildArgs, "test", "-destination", buildParams.DeviceDestination)
-	if buildTestParams.TestPlan != "" {
-		xcodebuildArgs = append(xcodebuildArgs, "-testPlan", buildTestParams.TestPlan)
+	if params.TestPlan != "" {
+		xcodebuildArgs = append(xcodebuildArgs, "-testPlan", params.TestPlan)
 	}
-	xcodebuildArgs = append(xcodebuildArgs, "-resultBundlePath", buildTestParams.TestOutputDir)
+	xcodebuildArgs = append(xcodebuildArgs, "-resultBundlePath", params.TestOutputDir)
 
-	if buildTestParams.GenerateCodeCoverage {
-		xcodebuildArgs = append(xcodebuildArgs, "GCC_INSTRUMENT_PROGRAM_FLOW_ARCS=YES")
-		xcodebuildArgs = append(xcodebuildArgs, "GCC_GENERATE_TEST_COVERAGE_FILES=YES")
+	if params.GenerateCodeCoverage {
+		xcodebuildArgs = append(xcodebuildArgs, "GCC_INSTRUMENT_PROGRAM_FLOW_ARCS=YES", "GCC_GENERATE_TEST_COVERAGE_FILES=YES")
 	}
 
-	if buildTestParams.AdditionalOptions != "" {
-		options, err := shellquote.Split(buildTestParams.AdditionalOptions)
+	if xcodeMajorVersion >= 13 && params.RetryTestsOnFailure {
+		xcodebuildArgs = append(xcodebuildArgs, "-retry-tests-on-failure")
+
+		// TODO(STEP-1054): Allow customization of `-test-iterations`.
+		xcodebuildArgs = append(xcodebuildArgs, "-test-iterations", "2")
+	}
+
+	if params.AdditionalOptions != "" {
+		options, err := shellquote.Split(params.AdditionalOptions)
 		if err != nil {
-			return "", 1, fmt.Errorf("failed to parse additional options (%s), error: %s", buildTestParams.AdditionalOptions, err)
+			return nil, fmt.Errorf("failed to parse additional options (%s), error: %s", params.AdditionalOptions, err)
 		}
 		xcodebuildArgs = append(xcodebuildArgs, options...)
 	}
 
-	xcprettyArgs := []string{}
-	if xcprettyOptions != "" {
-		options, err := shellquote.Split(xcprettyOptions)
-		if err != nil {
-			return "", 1, fmt.Errorf("failed to parse additional options (%s), error: %s", xcprettyOptions, err)
-		}
-		// get and delete the xcpretty output file, if exists
-		xcprettyOutputFilePath := ""
-		isNextOptOutputPth := false
-		for _, aOpt := range options {
-			if isNextOptOutputPth {
-				xcprettyOutputFilePath = aOpt
-				break
-			}
-			if aOpt == "--output" {
-				isNextOptOutputPth = true
-				continue
-			}
+	return xcodebuildArgs, nil
+}
+
+func handleTestRunError(prevRunParams testRunParams, prevRunResult testRunResult) (string, int, error) {
+	if prevRunParams.swiftPackagesPath != "" && isStringFoundInOutput(cache.SwiftPackagesStateInvalid, prevRunResult.xcodebuildLog) {
+		log.RWarnf("xcode-test", "swift-packages-cache-invalid", nil, "swift packages cache is in an invalid state")
+		if err := os.RemoveAll(prevRunParams.swiftPackagesPath); err != nil {
+			log.Errorf("failed to remove Swift package caches, error: %s", err)
+			return prevRunResult.xcodebuildLog, prevRunResult.exitCode, prevRunResult.err
 		}
```

Comment:
Should re-run in this case.

Reply:
This is the old version, which didn't rerun the tests either: https://github.com/bitrise-steplib/steps-xcode-test/pull/171/files/f197272d32e1fa0e03693942f4bd08de64d19c0e#diff-dccdd133d1820fcdb03d94b18332976edf6f3d73e9d73979a6d74c0bac37438bL146-L150.

I'm unfamiliar with the requirements/use cases here. Why should we rerun them?

Reply:
The original intent was to re-run them in case the Swift Package Cache gets in an invalid state (we restore the cache and it may be too Old or something like that). I changed it as there is no reason to keep the current buggy behaviour.